transport/tlsfrag/fixed_len_writer.go

```diff
@@ -0,0 +1,300 @@
+// Copyright 2024 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsfrag
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net"
+)
+
+// FixedLenFragFunc takes the length of the first [handshake record]'s content (without the 5-byte header),
+// and returns an integer that determines where the record should be fragmented.
+//
+// The returned splitLen should be in range 1 to recordLen-1.
+// The record content will then be fragmented into two parts: record[:splitLen] and record[splitLen:].
+// If splitLen is either ≤ 0 or ≥ recordLen, no fragmentation will occur.
+//
+// [handshake record]: https://datatracker.ietf.org/doc/html/rfc8446#section-5.1
+type FixedLenFragFunc func(recordLen int) (splitLen int)
+
+// fixedLenWriter splits the initial TLS Client Hello record into two TLS records based on a fixed length returned by
+// a [FixedLenFragFunc] callback.
+// These fragmented records are then written to the base [io.Writer]. Subsequent packets are not modified and are
+// directly transmitted through the base [io.Writer].
+type fixedLenWriter struct {
+	base io.Writer
+	frag FixedLenFragFunc
+	done bool   // the first fragmented header and payload are flushed (or don't split)
+	hdr  []byte // the 5 bytes header
+
+	// the records' sizes and written bytes (including 5 bytes header)
+	r1Size, r1Written, r2Size int
+}
+
+var _ io.Writer = (*fixedLenWriter)(nil)
+
+// NewFixedLenWriter creates a [io.Writer] that splits the first TLS Client Hello record into two records based
+// on the provided [FixedLenFragFunc] callback.
+// It then writes these records and all subsequent messages to the base [io.Writer].
+// If the first message isn't a Client Hello, no splitting occurs and all messages are written directly to base.
+//
+// The returned [io.Writer] will implement the [io.ReaderFrom] interface for optimized performance if the base
+// [io.Writer] implements [io.ReaderFrom].
+func NewFixedLenWriter(base io.Writer, frag FixedLenFragFunc) (io.Writer, error) {
```

Comment:
This is not fixed length. You use the length from the function. A "fixed length" writer would take a fixed value instead, which allows you to further optimize the code to write the incoming bytes as you receive them, without the need for buffering the record header and holding the initial packet.

Note that waiting for the record header may deadlock if all the client writes is less than 5 bytes (e.g. "PING"), so this implementation is still not safe to run on any port or connection.

That's ok for now, but consider an actual fixed implementation that provides that safety.

For this implementation, maybe name it something like NewRecordLenFuncWriter?

Reply:
Yes you are right. I can do `NewRecordLenFuncWriter`, similarly shall we change the `NewFixedLenStreamDialer` to `NewRecordLenStreamDialer` because it also accepts negative number, whose absolute value represents the record size of the second record?

But still I'm afraid we still need to store the packet header for a `FixedLenWriter`, because we are gonna use it as the second record's header (actually we only need the version, or we can always use `0x0301`).

Also, a deadlock similar to the `PING` scenario would still be a problem for a `FixedLenWriter`, because we need `p[3:5]` to determine whether it is a valid TLS handshake record (for example, `p[3:5] == 0` means it is invalid, we should not split). Therefore if we received `p = [0x16, 0x03, 0x01, 0x00]`, we can `Write(p[0:3])`, but we still need to wait for the fifth byte. If the request only contains 4 bytes, we will wait forever. Though this sequence is almost impossible in real network traffic.